_posts/2017-01-09-Jak-psat-php-Zavislosti+verzovaci-systemy+debugging.md

```diff
@@ -0,0 +1,38 @@
+
+Téměř každé IDE umožňuje nějaký stupeň integrace verzovacích systémů, neškodí se ale základy naučit přímo v příkazovém řádku a tedy univerzálně. Dobré návody a rozcestníky na git najdete třeba [tady](https://rogerdudler.github.io/git-guide/), [tady](https://stackoverflow.com/questions/315911/git-for-beginners-the-definitive-practical-guide) nebo [tady](https://www.sitepoint.com/git-for-beginners/). 
+
+# Testování a debugging
+Je dobrým zvykem mít kód otestovaný lépe než použitím `echo($foo)` a k tomu ručním proklikáním celého projektu v prohlížeči. To jsou první a intuitivní praktiky, kterými každý z nás prošel a používal, nesmíme na nich ale zamrznout! 
+
+Pro testování projektu se používají automatizované testy. Je třeba tyto testy napsat a je třeba mít aplikaci, která testy provádí. Testování se většinou dělí na Unit Testing (jednotkové testování), který testuje jednotlivé třidy nebo jiné malé celky a Behaviour/Acceptance Testing (testy chování), který testuje chování aplikace jako celku. 
+
+Pro Unit Testing se nejvíc používá [testovací framework PHPunit](https://phpunit.de/). Jako alternativy lze vyzkoušet mnohem jednodušší [SimpleTest](http://simpletest.org/) nebo český [tester](https://tester.nette.org/cs/) používaný ve frameworku Nette. Jako simulaci prohlížeče se pro Behaviour Testing používá [Selenium](http://www.seleniumhq.org/). Můžete zkusit i [Behat](http://behat.org/en/latest/) nebo [PHPSpec](http://www.phpspec.net/en/latest/). Pro komplexní testování lze využít třeba [Codeception](http://codeception.com/). 
+
+Na testování můžete mít postavený celý vývoj. Ti, co to mají zavedené, si to nemůžou vynachválit, je ale těžší pro začínající vývojáře. Tento způsob kódování se nazývá Test Driven Developement (TDD) nebo Behaviour Driven Developement (BDD). Obsáhlejší tutoriál najdete například [tady](https://code.tutsplus.com/tutorials/the-newbies-guide-to-test-driven-development--net-13835). 
+
+Vedle testování se pro ladění aplikace používají debuggery. V zásadě slouží k zastavení aplikace v libovolné fázi běhu a nahlédnutí do proměnných, objektů a podobně. Není potom potřeba rozbíjet výstup aplikace výpisy proměnných, které potřebujeme zkontrolovat. Pro PHP jsou rozšířené debuggery [Xdebug](https://xdebug.org/) (který používám), [Advanced PHP Debugger](https://pecl.php.net/package/apd) nebo [DBG degubber](http://www.php-debugger.com/dbg/). Debuggery jsou standardně implementované do vývojových prostředí a vyplatí se s nimi naučit dobře zacházet, protože umí řádově zrychlit ladění aplikace. 
```

Comment:
Osekal jsem, ale základní informace o debuggingu jsem ponechal - myslím že je důležité je pro představu zmínit